src/mysql_vm_helpers.py

```diff
@@ -271,6 +272,115 @@ def delete_temp_backup_directory(self) -> None:
             group=ROOT_SYSTEM_USER,
         )
 
+    def retrieve_backup_with_xbcloud(
+        self,
+        s3_bucket: str,
+        s3_path: str,
+        s3_access_key: str,
+        s3_secret_key: str,
+        backup_id: str,
+    ) -> Tuple[str, str, str]:
+        """Retrieve the provided backup with xbcloud."""
+        return super(MySQL, self).retrieve_backup_with_xbcloud(
+            s3_bucket,
+            s3_path,
+            s3_access_key,
+            s3_secret_key,
+            backup_id,
+            MYSQL_DATA_DIR,
+            CHARMED_MYSQL_XBCLOUD_LOCATION,
+            CHARMED_MYSQL_XBSTREAM_LOCATION,
+            user=ROOT_SYSTEM_USER,
+            group=ROOT_SYSTEM_USER,
+        )
+
+    def prepare_backup_for_restore(self, backup_location: str) -> Tuple[str, str]:
+        """Prepare the download backup for restore with xtrabackup --prepare."""
+        return super(MySQL, self).prepare_backup_for_restore(
+            backup_location,
+            CHARMED_MYSQL_XTRABACKUP_LOCATION,
+            XTRABACKUP_PLUGIN_DIR,
+            user=ROOT_SYSTEM_USER,
+            group=ROOT_SYSTEM_USER,
+        )
+
+    def empty_data_files(self) -> None:
+        """Empty the mysql data directory in preparation of the restore."""
+        super(MySQL, self).empty_data_files(
+            MYSQL_DATA_DIR,
+            user=ROOT_SYSTEM_USER,
+            group=ROOT_SYSTEM_USER,
+        )
+
+    def restore_backup(
+        self,
+        backup_location: str,
+    ) -> Tuple[str, str]:
+        """Restore the provided prepared backup."""
+        # TODO: remove workaround for changing permissions and ownership of data
+        # files once restore backup commands can be run with snap_daemon user
+        try:
+            # provide write permissions to root (group owner of the data directory)
+            # so the root user can move back files into the data directory
+            command = f"chmod 775 {MYSQL_DATA_DIR}".split()
+            subprocess.run(
+                command,
+                user=ROOT_SYSTEM_USER,
+                group=ROOT_SYSTEM_USER,
+                capture_output=True,
+                text=True,
+            )
+        except subprocess.CalledProcessError as e:
+            logger.exception("Failed to change data directory permissions before restoring")
+            raise MySQLRestoreBackupError(e)
+
+        stdout, stderr = super(MySQL, self).restore_backup(
+            backup_location,
+            CHARMED_MYSQL_XTRABACKUP_LOCATION,
+            MYSQLD_DEFAULTS_CONFIG_FILE,
+            MYSQL_DATA_DIR,
+            XTRABACKUP_PLUGIN_DIR,
+            user=ROOT_SYSTEM_USER,
+            group=ROOT_SYSTEM_USER,
+        )
+
+        try:
+            # Revert permissions for the data directory
+            command = f"chmod 755 {MYSQL_DATA_DIR}".split()
```

Comment:
can it work with 750?
```suggestion
            command = f"chmod 750 {MYSQL_DATA_DIR}".split()
```

Reply:
it does work with 750. changed in 023f6d77de1362d5d13154e2a4f11b95258d0d19